python/ray/air/util/tensor_extensions/pandas.py

```diff
@@ -703,28 +708,21 @@ def __init__(
 
         if isinstance(values, np.ndarray):
             if values.dtype.type is np.object_:
-                if len(values) == 0 or (
-                    not isinstance(values[0], str)
-                    and isinstance(
-                        values[0], (np.ndarray, TensorArrayElement, Sequence)
+                if len(values) == 0:
+                    # Tensor is empty, pass through to create empty TensorArray.
+                    pass
+                elif isinstance(values[0], str):
+                    # Tensor of strings, these don't need a TensorArray representation.
+                    raise TypeError(
+                        "Got a tensor of strings when constructing a TensorArray, but "
+                        "a tensor of strings doesn't need to be represented using a "
+                        "TensorArray."
                     )
-                ):
-                    # Convert ndarrays of ndarrays/TensorArrayElements
-                    # with an opaque object type to a properly typed ndarray of
-                    # ndarrays.
+                elif isinstance(values[0], (np.ndarray, TensorArrayElement, Sequence)):
+                    # Try to convert ndarrays of ndarrays/TensorArrayElements with an
+                    # opaque object type to a properly typed ndarray of ndarrays.
+                    # TODO (Clark): Maybe don't even try this conversion?
```

Comment:
Object-dtyped columns occasionally arise that _can_ be consolidated into a contiguous, well-typed ndarray, in which case we should do so since it makes everything far more efficient downstream.